src/components/Tag.ts

```diff
@@ -0,0 +1,151 @@
+import { Component, createElement } from "react";
+
+import * as TagsInput from "react-tagsinput";
+import { AutoComplete, Suggestion } from "./AutoComplete";
+import { Alert } from "./Alert";
+
+import * as classNames from "classnames";
+
+import "react-tagsinput/react-tagsinput.css";
+import "../ui/Tag.scss";
+
+export interface TagProps {
+    alertMessage?: string;
+    tagStyle?: BootstrapStyle;
+    className?: string;
+    createTag?: (tag: string) => void;
+    enableCreate?: boolean;
+    enableSuggestions?: boolean;
+    inputPlaceholder: string;
+    lazyLoad?: boolean;
+    lazyLoadTags?: () => void;
+    readOnly?: boolean;
+    showError: (message: string) => void;
+    style?: object;
+    tagLimit: number;
+    tagLimitMessage: string;
+    tags: string[];
+    tagValue: string;
+}
+
+interface TagState {
+    alertMessage?: string;
+    tag: string;
+    tags: string[];
+}
+
+export type BootstrapStyle = "primary" | "inverse" | "success" | "info" | "warning" | "danger";
+
+export class Tag extends Component<TagProps, TagState> {
+    static defaultProps: TagProps = {
+        tags: [],
+        inputPlaceholder: "Add a tag",
+        showError: () => undefined,
+        tagStyle: "primary",
+        tagLimit: 0,
+        tagLimitMessage: "",
+        tagValue: ""
+    };
+    constructor(props: TagProps) {
+        super(props);
+
+        this.state = {
+            alertMessage: props.alertMessage,
+            tag: this.props.tagValue,
+            tags: props.tags
+        };
+        this.addTag = this.addTag.bind(this);
```

Comment:
no need to bind a function that will be scoped correctly any way
`addTag: (tag: string) => this.addTag(tag),`
Think the others are still relevant

Reply:
bind removed